### PR TITLE
Fix phrases load, for backward compatibility

### LINKED ISSUE #1751: Gensim 3.1.0 Phrase loader incompatible with older models due to introduction of common_terms
I have a Phrases model that was computed using Gensim 2.2.0 but because of changes in 3.1.0 (I believe), I cannot load it anymore. This is the error I get:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<timed exec> in <module>()

~/Stuff/Sources/anaconda3/envs/nlp/lib/python3.6/site-packages/gensim/models/phrases.py in __init__(self, phrases_model)
    546         self.delimiter = phrases_model.delimiter
    547         self.scoring = phrases_model.scoring
--> 548         self.common_terms = phrases_model.common_terms
    549         corpus = self.pseudocorpus(phrases_model)
    550         self.phrasegrams = {}

AttributeError: 'Phrases' object has no attribute 'common_terms'
```

Cheers!

### PR DESCRIPTION
**Pull request by alexgarel**
This should fix #1751.
Also it gives backward compatible load for Phraser.

### CODE REVIEW

**Comment by alexgarel** [comment]
@menshikh-iv it's ready to validate.

File: gensim/models/phrases.py [hunk 1]
```diff
@@ -178,7 +178,61 @@ def analyze_sentence(self, sentence, threshold, common_terms, scorer):
...
+    @classmethod
+    def load(cls, *args, **kwargs):
+        """
+        Load a previously saved Phrases class. Handles backwards compatibility from
+            older Phrases versions which did not support  pluggable scoring functions.
+            Otherwise, relies on utils.load
+        """
+
+        # for python 2 and 3 compatibility. basestring is used to check if model.scoring is a string
+        try:
```
Comment (piskvorky) [inline gensim/models/phrases.py]: Gensim uses the `six` library to bridge py2/py3.

File: gensim/models/phrases.py [hunk 2]
```diff
@@ -178,7 +178,61 @@ def analyze_sentence(self, sentence, threshold, common_terms, scorer):
...
+                        cls.__name__)
+                    logger.info(
+                        'setting scoring method to npmi_scorer pluggable scoring method ' +
+                        'for compatibility')
+                    model.scoring = npmi_scorer
+                else:
+                    raise ValueError(
+                        'failed to load %s model with unknown scoring setting %s' %
+                        (cls.__name__, model.scoring))
+        # if there is non common_terms attribute, inizialize
```
Comment (piskvorky) [inline gensim/models/phrases.py]: typo - `initialize`

File: gensim/models/phrases.py (hunk 1 above)
Comment (alexgarel) [reply gensim/models/phrases.py]: ok, note that it was from a previous commit…

**Comment by alexgarel** [comment]
> What is the rationale behind introducing the new class (PhrasesTransformation)?
Phrases and Phraser needs the same `load` tweaks to preserve backward compatibility, and I wanted to avoid copy/pasting same code in both class (Phraser backward compatible load was not yet supported, but needs to).

**Comment by alexgarel** [comment]
@piskvorky fixed.

File: gensim/test/test_phrases.py [hunk 3]
```diff
@@ -419,9 +419,78 @@ def testSaveLoadNoScoring(self):
...
-        finally:
-            if os.path.exists("test_phrases_testSaveLoadNoScoring_temp_save.pkl"):
-                os.remove("test_phrases_testSaveLoadNoScoring_temp_save.pkl")
+    def testSaveLoadNoCommonTerms(self):
+        """ Saving and loading a Phrases objects without common_terms
+        This should ensure backwards compatibility with old versions of Phrases"""
+
+        with temporary_file("test.pkl") as fpath:
+            bigram = Phrases(self.sentences, min_count=1, threshold=1)
+            del(bigram.common_terms)
```
Comment (menshikh-iv) [inline gensim/test/test_phrases.py]: better solution - checkout to the previous version, tran+save model and try to load old model here (in your current implementation, we possibly skip other bugs, that can be hidden due to current bug)

File: gensim/models/phrases.py [hunk 4]
```diff
@@ -178,7 +178,55 @@ def analyze_sentence(self, sentence, threshold, common_terms, scorer):
...
+        # if no scoring parameter, use default scoring
+        if not hasattr(model, 'scoring'):
+            logger.info('older version of %s loaded without scoring function', cls.__name__)
+            logger.info('setting pluggable scoring method to original_scorer for compatibility')
+            model.scoring = original_scorer
+        # if there is a scoring parameter, and it's a text value, load the proper scoring function
+        if hasattr(model, 'scoring'):
+            if isinstance(model.scoring, six.string_types):
+                if model.scoring == 'default':
+                    logger.info(
```
Comment (menshikh-iv) [inline gensim/models/phrases.py]: nitpick: we use 140 char limits for code, no need to split this call into several lines

File: gensim/test/test_phrases.py [hunk 5]
```diff
@@ -8,17 +8,29 @@
...
 from gensim import utils
-from gensim.models.phrases import SentenceAnalyzer, Phrases, Phraser, pseudocorpus
+from gensim.models.phrases import SentenceAnalyzer, Phrases, Phraser
+from gensim.models.phrases import pseudocorpus, original_scorer
 from gensim.test.utils import common_texts
 
-if sys.version_info[0] >= 3:
-    unicode = str
+
+@contextlib.contextmanager
```
Comment (menshikh-iv) [inline gensim/test/test_phrases.py]: This looks useful, please move it to `gensim.test.utils` https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/test/utils.py

File: gensim/models/phrases.py (hunk 4 above)
Comment (alexgarel) [reply gensim/models/phrases.py]: argh travis failed, it was 120 not 140 @menshikh-iv, you joker ! ;-)

**Comment by alexgarel** [comment]
@menshikh-iv this should be ok.

Your proposal to test load backward compatibility was right, for I extend it to scoring and found there was a bug in it.

**Comment by menshikh-iv** [comment]
Big thanks for fast fix @alexgarel, I checked with 2.0.0, 1.0.1 - works fine!

File: gensim/models/phrases.py (hunk 4 above)
Comment (menshikh-iv) [reply gensim/models/phrases.py]: @alexgarel oh, sorry, my mistake (120 instead of 80, not 140)

**Pull request merged**